terminal: Add ctrl+click link detection with mouse movement

nihalxkumar opened the pull request:
Closes #41994

This PR introduces Element-bounded drag tolerance for Ctrl/Cmd+click in terminal. 

Previously, Ctrl/Cmd+click on terminal links required pixel-perfect accuracy. Any mouse movement during the click would cancel the navigation, making it frustrating to click on links, especially on high-DPI displays or with sensitive mice.

Users can now click anywhere within a clickable element (file path, URL, hyperlink), drag the cursor anywhere within that same element's boundaries and release to trigger navigation

Implementation:

  - Stores detected element metadata (`text` and `grid_range`) on Ctrl/Cmd+mouse-down
  - Tracks cursor position during drag, preserving click state while within element bounds
  - Verifies element match on mouse-up before triggering navigation
  - Uses existing `find_from_grid_point()` for element detection

Before:

[before.webm](https://github.com/user-attachments/assets/ee80de66-998e-4d8e-94d0-f5e65eb06d22)

After:

[after.webm](https://github.com/user-attachments/assets/7c9ddd9e-cfc1-4c79-b62c-78e9d909e6f4)

Release Notes:

- terminal: Fixed an issue where `ctrl|cmd+click` on links was very sensitive to mouse movement. Clicking links now tolerates mouse movement within the same clickable element, making link navigation more reliable

nihalxkumar committed:
feat(terminal): improve ctrl+click link detection with mouse movement

Track hyperlink on mouse down and verify same link on mouse up,
allowing ctrl+click to work even with slight mouse movement.

Closes #41994

nihalxkumar committed:
style: format terminal mouse handling code

Fix rustfmt violations in terminal ctrl+click implementation

nihalxkumar committed:
feat(terminal): clickable elements with ctrl+click and mouse movement

nihalxkumar committed:
refactor(terminal): simplify click detection per review

- Remove redundant mouse_down_link field and OSC-8 handler
- Simplify bounds check to use .contains()
- Unify all clickable elements through find_from_grid_point()

davewa commented on crates/terminal/src/terminal.rs:
```diff
@@ -831,6 +833,7 @@ pub struct Terminal {
     is_remote_terminal: bool,
     last_mouse_move_time: Instant,
     last_hyperlink_search_position: Option<Point<Pixels>>,
+    mouse_down_detected_element: Option<(String, Match)>,
```
Making this match the return type of `find_from_grid_point` simplifies the rest of the code, so `Option<(String, bool, Match)>`

davewa reviewed (commented):
Thanks for doing this work! I thought of a couple more changes that I think will improve/simplify this. Also, I think some of the comments just repeat the code, I'd prefer those be removed. Otherwise, looks good.

davewa reviewed (commented):
Also, I'd love to see a couple unit tests for this using `VisualTestContext::simulate_event()` 😎

nihalxkumar committed:
refactor(terminal): simplify implementation per review

- Rename mouse_down_detected_element to mouse_down_hyperlink
- Change type to Option<(String, bool, Match)> to match
find_from_grid_point
- Simplify overlap check to equality comparison
- Remove comments

davewa reviewed (commented):
You may have figured out that my review style is iterative. I hope you are having as much fun as I am, it is shaping up nicely! I noticed that the use of `FindHyperlink` here results in us always calling `find_from_grid_point` twice in a row on the same position, I added a suggestion there.

nihalxkumar committed:
terminal: hyperlink processing into separate method and add tests

nihalxkumar committed:
terminal: replace VisualTestContext with direct event handlers in tests and fix lock handling

nihalxkumar committed:
test(terminal): use secondary_key in hyperlink tests

davewa commented on crates/terminal/src/terminal.rs:
```diff
@@ -1876,6 +1915,31 @@ impl Terminal {
+                    position,
+                    self.last_content.terminal_bounds,
+                    self.last_content.display_offset,
+                );
```
I find it a little concerning that we now have two places that call `grid_point()` followed by `find_from_grid_point()` that differ in subtle ways that I don't fully understand the reasons for. I'm not really sure what should be done about this, but I don't feel great about it, as I suspect one or the other is incorrect.

davewa commented on crates/terminal/src/terminal.rs:
```diff
@@ -2847,4 +2914,205 @@ mod tests {
+                modifiers: Modifiers::secondary_key(),
+                click_count: 1,
+            };
+            terminal.mouse_up(&mouse_up, &*cx);
```
It appears that the re-borrows for `mouse_up` are unnecessary?

davewa commented on crates/terminal/src/terminal.rs:
```diff
@@ -1760,6 +1768,20 @@ impl Terminal {
     ) {
         let position = e.position - self.last_content.terminal_bounds.bounds.origin;
         if !self.mouse_mode(e.modifiers.shift) {
+            if let Some((_, _, element_range)) = &self.mouse_down_hyperlink {
```
This should be renamed to `hyperlink_range`

davewa reviewed (commented):
Looking really nice. I just added a few more comments, we're very close now 😎.

nihalxkumar committed:
refactor(terminal): simplify patterns, improve lock handling, extract test helpers

nihalxkumar commented on crates/terminal/src/terminal.rs:
```diff
@@ -1760,6 +1768,20 @@ impl Terminal {
     ) {
         let position = e.position - self.last_content.terminal_bounds.bounds.origin;
         if !self.mouse_mode(e.modifiers.shift) {
+            if let Some((_, _, element_range)) = &self.mouse_down_hyperlink {
```
fixed

nihalxkumar commented on crates/terminal/src/terminal.rs:
```diff
@@ -1876,6 +1915,31 @@ impl Terminal {
+                    &term_lock,
+                    point,
+                    &mut self.hyperlink_regex_searches,
+                ) {
```
fixed

nihalxkumar committed:
style(terminal): fix rustfmt formatting for test helper

davewa commented on crates/terminal/src/terminal.rs:
```diff
@@ -2847,4 +2992,72 @@ mod tests {
+        terminal.update(cx, |terminal, cx| {
+            let down_position = point(px(80.0), px(10.0));
+            let up_position = point(px(10.0), px(50.0));
+            let terminal_bounds = terminal.last_content.terminal_bounds.bounds;
```
This can move inside `ctrl_mouse_move_to()`

nihalxkumar replied:
done

davewa reviewed (commented):
One last improvement, then I'd say this is ready!

nihalxkumar committed:
refactor(terminal): move terminal_bounds inside ctrl_mouse_move_to helper

nihalxkumar commented:
I am limited by memory. I used to check every commit with `cargo build -j 2` but I stopped after `4242a75a162ad0a9809936ac04cc5a9df50c3c64`. Started a build now

nihalxkumar commented:
@probably-neb It's ready

probably-neb committed:
Merge branch 'main' into fix/terminal-ctrl-click-tolerance
